### PR TITLE
Ports TG cursed katana to replace our cursed katana.

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Removes current cursed katana.

Adds TG cursed katana, a force 15 weapon that can rapidly attack simple mobs (but not rapidly attack others), and has similar to martial arts, combos for special effects.

Is added as an implant, and is thus antidrop when active.
Retracting it without hitting someone (other than yourself) will hit you for 25 damage.

Has 50% block chance in melee, but not vs bullets.

In wizard spellbook for one point.

If the user dies with the cursed katana in them, the katana is ejected from them, and they dust.

Effects are from combos:

**Hilt Strike - Harm, Harm and Disarm**
22 damage from the success of the combo, combined of a 17 damage hit and 5.
Throws the target 5 tiles, confusing them, and if they are a ranged hostile mob, adds 2 seconds of ranged cooldown to them. Great for keeping something at bay.

**Wide Slice - Disarm, Harm and Harm**
20 damage to target, consisting of a hit for 15 and a hit for 5. Each mob within the tiles left, right, front, front left, and front right, get hit with a 15 damage slash. 

**Dash Attack - Harm, Disarm and Disarm**
17 damage hit, user slices 8 tiles forward, blocked by dense objects. Anyone in between is cut for 15 brute.

**Tendon Cut - Disarm, Disarm and Harm**
15 damage, applies a subtype of the saw bleed status effect. Effectively, if you do this move twice in a row, you get an extra 25 damage hit, no other effect. In optimal conditions, this is the best DPS option, however after 2 combos (6 hits) it deals 115 damage, equivalent to 8 combat knife hits, or 3 current katana hits.

**Dark Heal - Harm, Disarm, Harm and Disarm**
User lets the katana feast on their target (15 brute), and the katana heals the user... at a price. Heals the user for 45 brute / burn over 3 seconds. After, they take 3 brute every 3 seconds for 30 seconds. However, using a heal during this period refreshes the timer of brute damage, and increases it by one. So 30 brute, then 40, then 50...

**Shatter - Disarm, Harm, Disarm and Harm**
Huge hit, dealing 75 damage if the target is a hostile mob, 35 otherwise, and knocking down for 3 seconds. HOWEVER, this breaks the sword for 45 seconds.

## Why It's Good For The Game

Current cursed katana is problematic. A force 40 weapon, bulky or not, is kinda, not great for crew to have. It is for all intents and purposes, a better legal esword, barring the ease of storage.

This turns it into a strong cursed weapon, powerful in the right hands, but punishing if you don't hit anyone, great on hostile mobs, less great on humanoids. No ranged block, so disablers / lethals have no issues with it, but antidrop and melee block, so a disarm has nothing on you. It's damage is strong, but ignoring special effects, a fire axe will probably serve you better for humanoids.

If we get combat mode in the future, will be easier to use.

## Testing

Spawning in skrells and golaiths.
Spawning the katana organ, ensuring it implants right.
Confirming that all combos work (bar the beam effect)
Confirm damage seems right.
Confirm cooldown on ranged mobs works (it did too well, numbers had to be lowered)
Confirmed shattering works.

## Todo:

- [x] beam effect on charge
- [x] not letting it teleport on z1
- [x] checking over numbers more for balance.
- [x] PERHAPS make it banish you into the katana on death, so it has an actual curse to it other than oof owie on retract if not hitting anything?

## Changelog
:cl: Qwertytoforty Fikou
del: Old cursed katana
add: TG's cursed katana. An organ that is implanted into you on use, with powerfull combos, but a thirst for blood... examine it to see the combos.
/:cl:

### CODE REVIEW

File: code/__HELPERS/lists.dm [hunk 1]
```diff
@@ -900,3 +900,17 @@ proc/dd_sortedObjectList(list/incoming)
 	. = list()
 	for(var/thing in flat_list)
 		.[thing] = TRUE
+
+///compare two lists, returns TRUE if they are the same
+/proc/compare_list(list/l,list/d)
```
Comment (lewcc) [inline code/__HELPERS/lists.dm:905]:
```suggestion
/proc/compare_list(list/l, list/d)
```

File: code/__HELPERS/lists.dm [hunk 2]
```diff
@@ -900,3 +900,17 @@ proc/dd_sortedObjectList(list/incoming)
...
+	if(!islist(l) || !islist(d))
+		return FALSE
+
+	if(l.len != d.len)
+		return FALSE
+
+	for(var/i in 1 to l.len)
+		if(l[i] != d[i])
+			return FALSE
```
Comment (lewcc) [inline code/__HELPERS/lists.dm:914]:
```suggestion
	if(length(l) != length(d))
		return FALSE

	for(var/i in 1 to length(l))
		if(l[i] != d[i])
			return FALSE
```

File: code/datums/status_effects/buffs.dm [hunk 3]
```diff
@@ -91,8 +93,11 @@
 
 /datum/status_effect/void_price/tick()
 	playsound(owner, 'sound/weapons/bite.ogg', 50, 1)
-	owner.adjustBruteLoss(3)
+	owner.adjustBruteLoss(price)
 
+/datum/status_effect/void_price/refresh()
+	price++
+	return ..()
```
Comment (lewcc) [inline code/datums/status_effects/buffs.dm:100]:
```suggestion
	return ..()
	
```

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 4]
```diff
@@ -264,3 +264,278 @@
...
+/obj/item/cursed_katana/attack_self(mob/user)
+	. = ..()
+	reset_inputs(user, TRUE)
+
+/obj/item/cursed_katana/attack(mob/living/target, mob/user, click_parameters)
+	if(target.stat == DEAD || target == user) //No, you can not stab yourself to cloak / not take the penalty for not drawing blood
+		return ..()
+	if(HAS_TRAIT(user, TRAIT_PACIFISM))
+		to_chat(user, "<span class='warning'>You don't want to harm [target]!</span>")
+		return
```
Comment (lewcc) [inline code/modules/mining/lavaland/necropolis_chests.dm:389]: I think you want to return TRUE here to stop the attack chain
```suggestion
		return TRUE
```

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 5]
```diff
@@ -264,3 +264,278 @@
...
+
+	UnregisterSignal(source, COMSIG_MOVABLE_IMPACT)
+	source.apply_damage(5, BRUTE, "chest")
+	if(ishostile(source))
+		var/mob/living/simple_animal/hostile/target = source
+		target.ranged_cooldown += 2 SECONDS
+	else if(iscarbon(source))
+		var/mob/living/carbon/target = source
+		target.AdjustConfused(8 SECONDS)
+	return NONE
```
Comment (lewcc) [inline code/modules/mining/lavaland/necropolis_chests.dm:449]:
```suggestion
	return
```

**Review by lewcc — COMMENTED** [review]
Quick pass

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 4 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:389]: Chain stops without true

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 5 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:449]: Might be needed, confirming

Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:449]: Confirmed it is needed.

**Comment by Qwertytoforty** [comment]
> I'd rather see the cursed katana removed entirely than see it become just more useless tendril loot, which we already have too much of. A sharpened crusher/combat knife is easier to get, easier to use, deals more damage and won't dust you on death. Also, the fact that it's being made available to wizards just make the entire thing seem kinda...noob-baity? It'll already be somewhat noob-baity for miners, but there's no reason for a wizard to take it. I mean, you, yourself, admitted that a fire axe is better, and that's something a wizard has easy access to.

I would say it is far from useless. I said for humanoids a fire axe is probably better _ignoring special effects_ With special effects included, it still has use. Unlike the fire axe, it is also antidrop and one handed, with block chance. Yes, a crusher is better, but a crusher is a two handed bulky weapon, this is not. I would say this is far from useless loot wise, and while it may not be the highest damage item, for a wizard, it is a one point investment for an antidrop easy to store weapon. Compared to every other wizard weapon that you will want to take instant summons with, this is not bad for the price.

File: code/datums/status_effects/debuffs.dm [hunk 6]
```diff
@@ -137,6 +138,12 @@
 	else
 		new /obj/effect/temp_visual/bleed(get_turf(owner))
 
+/datum/status_effect/saw_bleed/bloodletting
+	id = "bloodletting"
+	bleed_cap = 7
+	bleed_damage = 25 //Seems weak (it is) but it also works on humans and bipasses armor SOOOO
```
Comment (Sirryan2002) [inline code/datums/status_effects/debuffs.dm:144]:
```suggestion
	bleed_damage = 25 //Seems weak (it is) but it also works on humans and bypasses armor SOOOO
```

File: code/game/gamemodes/wizard/spellbook.dm [hunk 7]
```diff
@@ -496,6 +496,14 @@
 	log_name = "SI"
 	category = "Weapons and Armors"
 
+/datum/spellbook_entry/item/spell_blade
```
Comment (Sirryan2002) [inline code/game/gamemodes/wizard/spellbook.dm:499]:
```suggestion
/datum/spellbook_entry/item/cursed_katana
```
this smells of untested code >:| (unless this somehow works in its current state?)

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 8]
```diff
@@ -264,3 +264,278 @@
...
 #undef RIGHT_HAND
 #undef LEFT_HAND
+
+/obj/item/organ/internal/cyberimp/arm/katana
+	name = "dark shard"
+	desc = "An eerie metal shard surrounded by dark energies."
+	icon = 'icons/obj/lavaland/artefacts.dmi'
+	icon_state = "cursed_katana_organ"
+	status = 0
+	contents = newlist(/obj/item/cursed_katana)
```
Comment (Sirryan2002) [inline code/modules/mining/lavaland/necropolis_chests.dm:274]: should probably add this item in the Initialize sequence

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 9]
```diff
@@ -264,3 +264,278 @@
...
+
+
+/obj/item/organ/internal/cyberimp/arm/katana/attack_self(mob/living/carbon/user, modifiers)
+	. = ..()
+	to_chat(user,"<span class='userdanger'>The mass goes up your arm and goes inside it!</span>")
```
Comment (Sirryan2002) [inline code/modules/mining/lavaland/necropolis_chests.dm:279]:
```suggestion
	to_chat(user,"<span class='userdanger'>The mass goes up your arm and inside it!</span>")
```

**Review by Sirryan2002 — COMMENTED** [review]
first pass

File: code/game/gamemodes/wizard/spellbook.dm (hunk 7 above)
Comment (Qwertytoforty) [reply code/game/gamemodes/wizard/spellbook.dm:499]: Whoops sorry

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 8 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:274]: Handled as all other arm implants are

File: code/datums/status_effects/debuffs.dm [hunk 10]
```diff
@@ -213,6 +220,23 @@
 /datum/status_effect/bluespace_slowdown/on_remove()
 	owner.next_move_modifier /= 2
 
+/datum/status_effect/katana_curse //tracking if user is dead
+	id = "katana_curse"
+	duration = -1
+	alert_type = null
+	var/deathTick = 0
```
Comment (hal9000PR) [inline code/datums/status_effects/debuffs.dm:227]: couldn't you use a signal for this instead? register a signal listening for `COMSIG_MOB_DEATH` when you implant the organ and remove it when you remove the organ.

File: code/game/gamemodes/wizard/spellbook.dm [hunk 11]
```diff
@@ -496,6 +496,14 @@
 	log_name = "SI"
 	category = "Weapons and Armors"
 
+/datum/spellbook_entry/item/cursed_katana
+	name = "Cursed Katana"
+	desc = "A cursed artifact, used to seal a horrible being inside the katana, which has now reformed. Can be used to make multiple powerful combos, examine it to see them. Can no be dropped. On death, you will dust."
```
Comment (hal9000PR) [inline code/game/gamemodes/wizard/spellbook.dm:501]:
```suggestion
	desc = "A cursed artefact, used to seal a horrible being inside the katana, which has now reformed. Can be used to make multiple powerful combos, examine it to see them. Can not be dropped. On death, you will dust."
```

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 12]
```diff
@@ -264,3 +264,278 @@
...
+			return FALSE
+	return ..()
+
+/obj/item/organ/internal/cyberimp/arm/katana/proc/user_death(mob/user)
+	Retract(TRUE)
+	user.visible_message("<span class='warning'>[user] begins to turn to dust, his soul being contained within [src]!</span>",
+		"<span class='userdanger'>You feel your body begin to turn to dust, your soul being drawn into [src]!</span>")
+	forceMove(get_turf(owner))
+	remove(user)
+	addtimer(CALLBACK(user, /mob/.proc/dust), 1 SECONDS)
```
Comment (hal9000PR) [inline code/modules/mining/lavaland/necropolis_chests.dm:314]: if you do this, whats the need for the status effect?

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 4 above)
Comment (hal9000PR) [reply code/modules/mining/lavaland/necropolis_chests.dm:389]: TRUE ensures after attack isnt called

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 13]
```diff
@@ -264,3 +264,278 @@
...
+	var/dir_to_target = get_dir(user_turf, get_turf(target))
+	var/static/list/cursed_katana_slice_angles = list(0, -45, 45, -90, 90) //so that the animation animates towards the target clicked and not towards a side target
+	for(var/iteration in cursed_katana_slice_angles)
+		var/turf/turf = get_step(user_turf, turn(dir_to_target, iteration))
+		user.do_attack_animation(turf, ATTACK_EFFECT_CLAW)
+		for(var/mob/living/additional_target in turf)
+			if(user.Adjacent(additional_target) && additional_target.density)
+				additional_target.apply_damage(15, BRUTE, "chest", TRUE)
+				to_chat(additional_target, "<span class='userdanger'>You've been sliced by [user]!</span>")
+	target.apply_damage(5, BRUTE, "chest", TRUE)
```
Comment (hal9000PR) [inline code/modules/mining/lavaland/necropolis_chests.dm:465]: I would prefer we not name variables `turf`

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/debuffs.dm (hunk 10 above)
Comment (Qwertytoforty) [reply code/datums/status_effects/debuffs.dm:227]: Signal was mentioned on discord, did not quite work right, and would not pass cl due to sleeps in CL

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 13 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:465]: can be renamed

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 12 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:314]: See signal issues

Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:314]: nevermind I am dumb, duplicate

Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:314]: I am twice as dumb and should not read code I wrote 2 weeks ago after waking up. Called by the debuff

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 4 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm]: done

File: code/datums/status_effects/debuffs.dm (hunk 6 above)
Comment (Qwertytoforty) [reply code/datums/status_effects/debuffs.dm]: fixed

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 13 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm]: renamed

File: code/datums/status_effects/debuffs.dm (hunk 10 above)
Comment (Qwertytoforty) [reply code/datums/status_effects/debuffs.dm]: signal now working

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 14]
```diff
@@ -264,3 +264,284 @@
...
+	var/dir_to_target = get_dir(user_turf, get_turf(target))
+	var/static/list/cursed_katana_slice_angles = list(0, -45, 45, -90, 90) //so that the animation animates towards the target clicked and not towards a side target
+	for(var/iteration in cursed_katana_slice_angles)
+		var/turf/T = get_step(user_turf, turn(dir_to_target, iteration))
+		user.do_attack_animation(T, ATTACK_EFFECT_CLAW)
+		for(var/mob/living/additional_target in T)
+			if(user.Adjacent(additional_target) && additional_target.density)
+				additional_target.apply_damage(15, BRUTE, "chest", TRUE)
+				to_chat(additional_target, "<span class='userdanger'>You've been sliced by [user]!</span>")
+	target.apply_damage(5, BRUTE, "chest", TRUE)
```
Comment (hal9000PR) [inline code/modules/mining/lavaland/necropolis_chests.dm:471]: would you use the body zone definitions for these procs?

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 15]
```diff
@@ -264,3 +264,284 @@
...
+
+/obj/item/cursed_katana/proc/heal(mob/living/target, mob/living/user)
+	user.visible_message("<span class='warning'>[user] lets [src] feast on [target]'s blood!</span>",
+		"<span class='warning'>You let [src] feast on [target], and it heals you, at a price!</span>")
+	target.apply_damage(15, BRUTE, "chest", TRUE)
```
Comment (hal9000PR) [inline code/modules/mining/lavaland/necropolis_chests.dm:476]: and here

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 16]
```diff
@@ -264,3 +264,284 @@
...
+	target.apply_damage(17, BRUTE, "chest", TRUE)
+	for(var/distance in 0 to 8)
+		var/turf/current_dash_target = dash_target
+		current_dash_target = get_step(current_dash_target, user.dir)
+		if(!is_blocked_turf(current_dash_target, TRUE))
+			dash_target = current_dash_target
+		else
+			break
+		for(var/mob/living/additional_target in dash_target) //Slash through every mob you cut through
+			additional_target.apply_damage(15, BRUTE, "chest", TRUE)
```
Comment (hal9000PR) [inline code/modules/mining/lavaland/necropolis_chests.dm:511]: and here

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 17]
```diff
@@ -264,3 +264,284 @@
...
+			to_chat(additional_target, "<span class='userdanger'>You've been sliced by [user]!</span>")
+	user_turf.Beam(dash_target, icon_state = "warp_beam", time = 0.3 SECONDS, maxdistance = INFINITY)
+	user.forceMove(dash_target)
+
+/obj/item/cursed_katana/proc/shatter(mob/living/target, mob/user)
+	user.visible_message("<span class='warning'>[user] shatters [src] over [target]!</span>",
+		"<span class='notice'>You shatter [src] over [target]!</span>")
+	to_chat(target, "<span class='userdanger'>[user] shatters [src] over you!</span>")
+	target.apply_damage((ishostile(target) ? 75 : 35), BRUTE, "chest", TRUE)
```
Comment (hal9000PR) [inline code/modules/mining/lavaland/necropolis_chests.dm:520]: and here

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 18]
```diff
@@ -264,3 +264,284 @@
...
+	icon_state = "cursed_katana_organ"
+	status = 0
+	contents = newlist(/obj/item/cursed_katana)
+
+
+/obj/item/organ/internal/cyberimp/arm/katana/attack_self(mob/living/carbon/user, modifiers)
+	. = ..()
+	to_chat(user,"<span class='userdanger'>The mass goes up your arm and inside it!</span>")
+	playsound(user, 'sound/misc/demon_consume.ogg', 50, TRUE)
+	//user.apply_status_effect(STATUS_EFFECT_KATANA_CURSE)
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:281]: whoops, can do

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 14 above)
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm]: done

File: code/modules/mining/lavaland/necropolis_chests.dm (hunk 18 above)
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm]: removed

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 19]
```diff
@@ -264,3 +264,284 @@
...
+		var/list/combo_specifics = combo_list[combo]
+		if(compare_list(input_list,combo_specifics[COMBO_STEPS]))
+			INVOKE_ASYNC(src, combo_specifics[COMBO_PROC], target, user)
+			return TRUE
+	return FALSE
+
+/obj/item/cursed_katana/proc/reset_inputs(mob/user, deltimer)
+	input_list.Cut()
+	if(user)
+		to_chat(user, "<span class='notice'>You return to neutral stance</span>")
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm]: done

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 20]
```diff
@@ -264,3 +264,284 @@
...
+	if(deltimer && timerid)
+		deltimer(timerid)
+
+/obj/item/cursed_katana/proc/strike(mob/living/target, mob/user)
+	user.visible_message("<span class='warning'>[user] strikes [target] with [src]'s hilt!</span>",
+		"<span class='notice'>You hilt strike [target]!</span>")
+	to_chat(target, "<span class='userdanger'>You've been struck by [user]!</span>")
+	playsound(src, 'sound/weapons/genhit3.ogg', 50, TRUE)
+	RegisterSignal(target, COMSIG_MOVABLE_IMPACT, .proc/strike_throw_impact)
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/modules/mining/lavaland/necropolis_chests.dm:436]: now removes automatically after 5 seconds

**Review by hal9000PR — APPROVED** [review]

**Pull request merged**